### PR TITLE
[WIP] refactor train

### PR DESCRIPTION
**Pull request by TingquanGao**
TODO:
1. 优化AMP；部分完成；
2. 优化build_dataloader；shengyu；
3. 优化train_epoch；done；
4. 验证跑通shitu；

### CODE REVIEW

File: ppcls/arch/__init__.py [hunk 1]
```diff
@@ -35,20 +35,28 @@ def build_model(config, mode="train"):
     arch_config = copy.deepcopy(config["Arch"])
     model_type = arch_config.pop("name")
     use_sync_bn = arch_config.pop("use_sync_bn", False)
-    mod = importlib.import_module(__name__)
-    arch = getattr(mod, model_type)(**arch_config)
+
+    if hasattr(backbone_zoo, model_type):
```
Comment (TingquanGao) [inline ppcls/arch/__init__.py]: arch/global.task = class/rec/xxx

File: ppcls/arch/__init__.py [hunk 2]
```diff
@@ -65,12 +73,29 @@ def apply_to_static(config, model):
     return model
 
 
+# TODO(gaotingquan): export model
+class ClassModel(TheseusLayer):
```
Comment (TingquanGao) [inline ppcls/arch/__init__.py]: 去掉这一层

使用装饰器完成amp的功能

File: ppcls/arch/__init__.py [hunk 3]
```diff
@@ -65,12 +73,29 @@ def apply_to_static(config, model):
...
+        return self.backbone(x)
+
+
 class RecModel(TheseusLayer):
     def __init__(self, **config):
         super().__init__()
         backbone_config = config["Backbone"]
         backbone_name = backbone_config.pop("name")
-        self.backbone = eval(backbone_name)(**backbone_config)
+        self.backbone = getattr(backbone_zoo, backbone_name)(**backbone_config)
```
Comment (TingquanGao) [inline ppcls/arch/__init__.py]: backbone_zoo -> backbone

File: ppcls/arch/__init__.py (hunk 2 above)
Comment (TingquanGao) [reply ppcls/arch/__init__.py]: 先测试装饰器在动转静、静态图下

File: ppcls/data/__init__.py [hunk 4]
```diff
@@ -164,7 +166,7 @@ def mix_collate_fn(batch):
     if batch_sampler is None:
         data_loader = DataLoader(
             dataset=dataset,
-            places=device,
+            places=paddle.device.get_device(),
```
Comment (TingquanGao) [inline ppcls/data/__init__.py]: device == paddle.device.get_device() ？

File: ppcls/data/__init__.py [hunk 5]
```diff
@@ -176,13 +178,103 @@ def mix_collate_fn(batch):
...
             use_shared_memory=use_shared_memory,
             batch_sampler=batch_sampler,
             collate_fn=batch_collate_fn,
             worker_init_fn=init_fn)
 
+    total_samples = len(
+        data_loader.dataset) if not use_dali else data_loader.size
+    max_iter = len(data_loader) - 1 if platform.system() == "Windows" else len(
+        data_loader)
+    data_loader.max_iter = max_iter
```
Comment (TingquanGao) [inline ppcls/data/__init__.py:199]: data_loader 可能有这个属性

File: ppcls/data/__init__.py [hunk 6]
```diff
@@ -176,13 +178,103 @@ def mix_collate_fn(batch):
...
+        except Exception:
+            # NOTE: reset DALI dataloader manually
+            if self.use_dali:
+                self.dataloader.reset()
+            self.iterator = iter(self.dataloader)
+            batch = next(self.iterator)
+        return batch
+
+
+def build_dataloader(engine):
```
Comment (TingquanGao) [inline ppcls/data/__init__.py]: 传config

File: ppcls/data/__init__.py [hunk 7]
```diff
@@ -176,13 +178,103 @@ def mix_collate_fn(batch):
...
+    if "class_num" in engine.config["Global"]:
```
Comment (TingquanGao) [inline ppcls/data/__init__.py]: 兼容性代码删除掉

File: ppcls/data/__init__.py [hunk 8]
```diff
@@ -176,13 +178,103 @@ def mix_collate_fn(batch):
...
+        self.use_dali = use_dali
+        self.iterator = iter(dataloader)
+        self.max_iter = dataloader.max_iter
+        self.total_samples = dataloader.total_samples
+
+    def get_batch(self):
+        # fetch data batch from dataloader
+        try:
+            batch = next(self.iterator)
+        except Exception:
```
Comment (TingquanGao) [inline ppcls/data/__init__.py]: dataloader 长度 放在dataset实现

File: ppcls/engine/engine.py [hunk 9]
```diff
@@ -52,321 +51,69 @@ def __init__(self, config, mode="train"):
...
         # set device
-        assert self.config["Global"][
-            "device"] in ["cpu", "gpu", "xpu", "npu", "mlu", "ascend"]
-        self.device = paddle.set_device(self.config["Global"]["device"])
-        logger.info('train with paddle {} and device {}'.format(
-            paddle.__version__, self.device))
+        self._init_device()
 
         # gradient accumulation
         self.update_freq = self.config["Global"].get("update_freq", 1)
```
Comment (TingquanGao) [inline ppcls/engine/engine.py]: 尽量不保留  self.update_freq

File: ppcls/engine/engine.py [hunk 10]
```diff
@@ -52,321 +51,69 @@ def __init__(self, config, mode="train"):
...
-            loss_config = self.config.get("Loss", None)
-            if loss_config is not None:
-                loss_config = loss_config.get("Eval")
-                if loss_config is not None:
-                    self.eval_loss_func = build_loss(loss_config)
-                else:
-                    self.eval_loss_func = None
-            else:
-                self.eval_loss_func = None
+        self.train_loss_func, self.unlabel_train_loss_func, self.eval_loss_func = build_loss(
```
Comment (TingquanGao) [inline ppcls/engine/engine.py]: 改成 loss_dict

File: ppcls/engine/engine.py [hunk 11]
```diff
@@ -52,321 +51,69 @@ def __init__(self, config, mode="train"):
...
-            random.seed(int(seed) + dist.get_rank())
+        self._init_dist()
 
-        # build postprocess for infer
-        if self.mode == 'infer':
-            self.preprocess_func = create_operators(self.config["Infer"][
-                "transforms"])
-            self.postprocess_func = build_postprocess(self.config["Infer"][
-                "PostProcess"])
+        print_config(config)
```
Comment (TingquanGao) [inline ppcls/engine/engine.py]: 确保config是传引用

File: ppcls/engine/engine.py [hunk 12]
```diff
@@ -376,33 +123,35 @@ def train(self):
...
-        self.global_step = 0
 
-        if self.config.Global.checkpoints is not None:
-            metric_info = init_model(self.config.Global, self.model,
-                                     self.optimizer, self.train_loss_func,
-                                     ema_module)
-            if metric_info is not None:
-                best_metric.update(metric_info)
+        # build EMA model
+        self.model_ema = self._build_ema_model()
```
Comment (TingquanGao) [inline ppcls/engine/engine.py]: 是否可以放在build model

File: ppcls/engine/engine.py [hunk 13]
```diff
@@ -376,33 +123,35 @@ def train(self):
...
+            opt_name="optimizer",
+            model_ema_name="model_ema")
+
+        self._init_checkpoints(best_metric)
+
+        # global iter counter
+        self.global_step = 0
+        for epoch_id in range(best_metric["epoch"] + 1, epochs + 1):
             # for one epoch train
             self.train_epoch_func(self, epoch_id, print_batch_step)
```
Comment (TingquanGao) [inline ppcls/engine/engine.py]: train_epoch_func 改成类

File: ppcls/engine/engine.py [hunk 14]
```diff
@@ -376,33 +123,35 @@ def train(self):
...
+        # build EMA model
+        self.model_ema = self._build_ema_model()
+        # TODO: mv best_metric_ema to best_metric dict
+        best_metric_ema = 0
 
-        for epoch_id in range(best_metric["epoch"] + 1,
-                              self.config["Global"]["epochs"] + 1):
-            acc = 0.0
+        # build model saver
+        model_saver = ModelSaver(
```
Comment (TingquanGao) [inline ppcls/engine/engine.py]: 放在__init__()中

File: ppcls/engine/engine.py [hunk 15]
```diff
@@ -376,33 +123,35 @@ def train(self):
...
+            self,
```
Comment (TingquanGao) [inline ppcls/engine/engine.py]: 尽量不用self

File: ppcls/engine/engine.py [hunk 16]
```diff
@@ -52,321 +51,69 @@ def __init__(self, config, mode="train"):
...
+        print_config(config)
 
     def train(self):
         assert self.mode == "train"
         print_batch_step = self.config['Global']['print_batch_step']
         save_interval = self.config["Global"]["save_interval"]
+        start_eval_epoch = self.config["Global"].get("start_eval_epoch", 0) - 1
+        epochs = self.config["Global"]["epochs"]
+
         best_metric = {
```
Comment (TingquanGao) [inline ppcls/engine/engine.py]: 放在构造函数中

**Review by TingquanGao — COMMENTED** [review]
1. metric 封装类；
2. amp 分别放在model 和 loss中，不在train中；done；
3. ReduceOnPlateau 根据配置文件决定epoch/step；
4. engine去掉train方法；done；
5. 写Trainer：done；
```python
class trainer
 def train
 def train_epoch
```
6. 装饰器解决模型amp的问题；部分完成；

File: ppcls/arch/__init__.py (hunk 2 above)
Comment (TingquanGao) [reply ppcls/arch/__init__.py]: amp 装饰器实现部分完成；

**Review by weisy11 — APPROVED** [review]
LGTM
